NuGet.Client repo introduced Microsoft.Web.XDT prebuilts during .NET 7.0.200 source-build

### NuGet Product Used

Other/NA

### Product Version

nuget.client commit 20be995d651808feb5cd48f30cb97197e497377c

### Worked before?

c0de924b0cfc8089f4ce1f35360cb9cddcd81138

### Impact

It's more difficult to complete my work

### Repro Steps & Context

Build failure: https://github.com/dotnet/installer/pull/15018

Context: NuGet.Client is included in the .NET SDK. When building the SDK from source for Linux, we cannot download packages from the internet that we don't produce during our build. Microsoft.Web.XDT is consumed by NuGet.Client and also produced during a .NET build from source. We rely on using the property `MicrosoftWebXdtPackageVersion` to override the version used in this repo with the one we need at build-time. The change https://github.com/NuGet/NuGet.Client/commit/20be995d651808feb5cd48f30cb97197e497377c resulted in using a hardcoded version of 3.0.0. Instead, we should use the version `MicrosoftWebXdtPackageVersion` if it exists, otherwise use version 3.0.0 (or whatever other version this repo might need).

We can work around this right now using a patch, but this will need to be fixed as it is a regression ever since NuGet.Client adopted central package version management.

### Verbose Logs

```shell
/tarball/src/nuget-client/src/NuGet.Core/NuGet.PackageManagement/NuGet.PackageManagement.csproj : error NU1103: Unable to find a stable package Microsoft.Web.Xdt with version (>= 3.0.0) [/tarball/src/nuget-client/build/restorehelper.targets]
  /tarball/src/nuget-client/src/NuGet.Core/NuGet.PackageManagement/NuGet.PackageManagement.csproj : error NU1103:   - Found 1 version(s) in source-built [ Nearest version: 7.0.0-preview.22423.2 ] [/tarball/src/nuget-client/build/restorehelper.targets]
  /tarball/src/nuget-client/src/NuGet.Core/NuGet.PackageManagement/NuGet.PackageManagement.csproj : error NU1103:   - Found 1 version(s) in previously-source-built [ Nearest version: 7.0.0-preview.22423.2 ] [/tarball/src/nuget-client/build/restorehelper.targets]
  /tarball/src/nuget-client/src/NuGet.Core/NuGet.PackageManagement/NuGet.PackageManagement.csproj : error NU1103:   - Found 0 version(s) in prebuilt [/tarball/src/nuget-client/build/restorehelper.targets]
  /tarball/src/nuget-client/src/NuGet.Core/NuGet.PackageManagement/NuGet.PackageManagement.csproj : error NU1103:   - Found 0 version(s) in reference-packages [/tarball/src/nuget-client/build/restorehelper.targets]
  /tarball/src/nuget-client/src/NuGet.Core/NuGet.Build.Tasks/NuGet.Build.Tasks.csproj : error NU1103: Unable to find a stable package Microsoft.Web.Xdt with version (>= 3.0.0) [/tarball/src/nuget-client/build/restorehelper.targets]
  /tarball/src/nuget-client/src/NuGet.Core/NuGet.Build.Tasks/NuGet.Build.Tasks.csproj : error NU1103:   - Found 1 version(s) in source-built [ Nearest version: 7.0.0-preview.22423.2 ] [/tarball/src/nuget-client/build/restorehelper.targets]
  /tarball/src/nuget-client/src/NuGet.Core/NuGet.Build.Tasks/NuGet.Build.Tasks.csproj : error NU1103:   - Found 1 version(s) in previously-source-built [ Nearest version: 7.0.0-preview.22423.2 ] [/tarball/src/nuget-client/build/restorehelper.targets]
  /tarball/src/nuget-client/src/NuGet.Core/NuGet.Build.Tasks/NuGet.Build.Tasks.csproj : error NU1103:   - Found 0 version(s) in prebuilt [/tarball/src/nuget-client/build/restorehelper.targets]
  /tarball/src/nuget-client/src/NuGet.Core/NuGet.Build.Tasks/NuGet.Build.Tasks.csproj : error NU1103:   - Found 0 version(s) in reference-packages [/tarball/src/nuget-client/build/restorehelper.targets]
```

> @jeffkl: Is there a definitive list of package properties that a repo must have?

Unfortunately, I don't have an exhaustive list for you. Any version properties introduced by @crummel should probably stay. I'm not seeing any more failures to restore after making the following change: https://github.com/dotnet/installer/pull/15018/commits/d238d9b6b6ffbc73ab200413d82656824edf5571.

NuGet.Client still fails to build after that change with the following error though:
```
    ...
    NuGet.PackageManagement -> /data/tb/15018/src/nuget-client/artifacts/NuGet.PackageManagement/bin/net7.0/NuGet.PackageManagement.dll
    NuGet.Packaging.Core -> /data/tb/15018/src/nuget-client/artifacts/NuGet.Packaging.Core/bin/net472/NuGet.Packaging.Core.dll
    NuGet.Packaging.Core -> /data/tb/15018/src/nuget-client/artifacts/NuGet.Packaging.Core/bin/netstandard2.0/NuGet.Packaging.Core.dll
    NuGet.Packaging.Core -> /data/tb/15018/src/nuget-client/artifacts/NuGet.Packaging.Core/bin/net7.0/NuGet.Packaging.Core.dll
  /data/tb/15018/src/nuget-client/build/build.proj(176,5): error MSB4181: The "MSBuild" task returned false but did not log an error.

  Build FAILED.

  CSC : error CS9057: The analyzer assembly '/data/tb/15018/packages/restored/system.text.json/7.0.0/analyzers/dotnet/roslyn4.4/cs/System.Text.Json.SourceGeneration.dll' references version '4.5.0.0' of the compiler, which is newer than the currently running version '4.4.0.0'. [/data/tb/15018/src/nuget-client/src/NuGet.Core/NuGet.Build.Tasks.Console/NuGet.Build.Tasks.Console.csproj::TargetFramework=net7.0]
  CSC : error CS9057: The analyzer assembly '/data/tb/15018/packages/restored/system.text.json/7.0.0/analyzers/dotnet/roslyn4.4/cs/System.Text.Json.SourceGeneration.dll' references version '4.5.0.0' of the compiler, which is newer than the currently running version '4.4.0.0'. [/data/tb/15018/src/nuget-client/src/NuGet.Core/NuGet.CommandLine.XPlat/NuGet.CommandLine.XPlat.csproj::TargetFramework=net7.0]
  /data/tb/15018/src/nuget-client/build/build.proj(176,5): error MSB4181: The "MSBuild" task returned false but did not log an error.
      0 Warning(s)
      3 Error(s)
```

Do you think this could be package version related? It does look like you also touched the `Microsoft.Net.Compilers`. I'll try out a fix for that tomorrow morning.

--- a/Directory.Packages.props
+++ b/Directory.Packages.props
@@ -2,21 +2,21 @@
   <PropertyGroup>
     <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
     
-    <CryptographyPackagesVersion>5.0.0</CryptographyPackagesVersion>
-    <MicrosoftCodeAnalysisPublicApiAnalyzersVersion>3.0.0</MicrosoftCodeAnalysisPublicApiAnalyzersVersion>
-    <NewtonsoftJsonPackageVersion>13.0.1</NewtonsoftJsonPackageVersion>
-    <SystemPackagesVersion>4.3.0</SystemPackagesVersion>
-    <VSFrameworkVersion>17.2.0-preview-2-32304-091</VSFrameworkVersion>
-    <VSServicesVersion>16.153.0</VSServicesVersion>
+    <CryptographyPackagesVersion Condition="'$(CryptographyPackagesVersion)' == ''">5.0.0</CryptographyPackagesVersion>
+    <MicrosoftCodeAnalysisPublicApiAnalyzersVersion Condition="'$(MicrosoftCodeAnalysisPublicApiAnalyzersVersion)' == ''">3.0.0</MicrosoftCodeAnalysisPublicApiAnalyzersVersion>
+    <NewtonsoftJsonPackageVersion Condition="'$(NewtonsoftJsonPackageVersion)' == ''">13.0.1</NewtonsoftJsonPackageVersion>
+    <SystemPackagesVersion Condition="'$(SystemPackagesVersion)' == ''">4.3.0</SystemPackagesVersion>
+    <VSFrameworkVersion Condition="'$(VSFrameworkVersion)' == ''">17.2.0-preview-2-32304-091</VSFrameworkVersion>
+    <VSServicesVersion Condition="'$(VSServicesVersion)' == ''">16.153.0</VSServicesVersion>
     <!-- Default MSBuild version -->
-    <MicrosoftBuildVersion>17.3.1</MicrosoftBuildVersion>
+    <MicrosoftBuildVersion Condition="'$(MicrosoftBuildVersion)' == ''">17.3.1</MicrosoftBuildVersion>
     <!-- The last package version of MSBuild that works with netcoreapp3.1 or netstandard2.0 is 16.8.0.  Our assemblies are still compiled against MSBuild assembly version 15.1.0.0 which will work with all versions -->
     <MicrosoftBuildVersion Condition="'$(TargetFramework)' == 'netcoreapp3.1' Or '$(TargetFramework)' == 'netstandard2.0'">16.8.0</MicrosoftBuildVersion>
     <!-- The last package version of MSBuild that works with net5.0 is 16.11.0.  Our assemblies are still compiled against MSBuild assembly version 15.1.0.0 which will work with all versions -->
     <MicrosoftBuildVersion Condition="'$(TargetFramework)' == 'netcoreapp5.0'">16.11.0</MicrosoftBuildVersion>
     
     <!-- Overridden by source build to ensure the same version is used across products, do not remove this property -->
-    <MicrosoftWebXdtPackageVersion>3.0.0</MicrosoftWebXdtPackageVersion>
+    <MicrosoftWebXdtPackageVersion Condition="'$(MicrosoftWebXdtPackageVersion)' == ''">3.0.0</MicrosoftWebXdtPackageVersion>
   </PropertyGroup>
 
   <ItemGroup>

--- a/Directory.Packages.props
+++ b/Directory.Packages.props
@@ -15,8 +15,9 @@
     <!-- The last package version of MSBuild that works with net5.0 is 16.11.0.  Our assemblies are still compiled against MSBuild assembly version 15.1.0.0 which will work with all versions -->
     <MicrosoftBuildVersion Condition="'$(TargetFramework)' == 'netcoreapp5.0'">16.11.0</MicrosoftBuildVersion>
     
-    <!-- Overridden by source build to ensure the same version is used across products, do not remove this property -->
+    <!-- Overridden by source build to ensure the same version is used across products, do not remove these properties -->
     <MicrosoftWebXdtPackageVersion Condition="'$(MicrosoftWebXdtPackageVersion)' == ''">3.0.0</MicrosoftWebXdtPackageVersion>
+    <SystemComponentModelCompositionPackageVersion Condition="'$(SystemComponentModelCompositionPackageVersion)' == ''">4.5.0</SystemComponentModelCompositionPackageVersion>
   </PropertyGroup>
 
   <ItemGroup>
@@ -71,7 +72,7 @@
     <PackageVersion Include="SharpZipLib" Version="1.3.3" />
     <PackageVersion Include="System.Collections" Version="$(SystemPackagesVersion)" />
     <PackageVersion Include="System.CommandLine" Version="2.0.0-beta4.22272.1" />
-    <PackageVersion Include="System.ComponentModel.Composition" Version="4.5.0" />
+    <PackageVersion Include="System.ComponentModel.Composition" Version="$(SystemComponentModelCompositionPackageVersion)" />
     <PackageVersion Include="System.Configuration.ConfigurationManager" Version="4.4.0" />
     <PackageVersion Include="System.Diagnostics.Debug" Version="$(SystemPackagesVersion)" />
     <PackageVersion Include="System.IO.FileSystem.Primitives" Version="$(SystemPackagesVersion)" />

--- a/Directory.Packages.props
+++ b/Directory.Packages.props
@@ -14,6 +14,9 @@
     <MicrosoftBuildVersion Condition="'$(TargetFramework)' == 'netcoreapp3.1' Or '$(TargetFramework)' == 'netstandard2.0'">16.8.0</MicrosoftBuildVersion>
     <!-- The last package version of MSBuild that works with net5.0 is 16.11.0.  Our assemblies are still compiled against MSBuild assembly version 15.1.0.0 which will work with all versions -->
     <MicrosoftBuildVersion Condition="'$(TargetFramework)' == 'netcoreapp5.0'">16.11.0</MicrosoftBuildVersion>
+    
+    <!-- Overridden by source build to ensure the same version is used across products, do not remove this property -->
+    <MicrosoftWebXdtPackageVersion>3.0.0</MicrosoftWebXdtPackageVersion>
   </PropertyGroup>
 
   <ItemGroup>
@@ -58,7 +61,7 @@
     <PackageVersion Include="Microsoft.VisualStudio.VCProjectEngine" Version="$(VSFrameworkVersion)" />
     <PackageVersion Include="Microsoft.VisualStudio.Workspace.VSIntegration" Version="17.0.7-preview-0001-g5492e466a9" />
     <PackageVersion Include="Microsoft.VSSDK.BuildTools" Version="17.0.1600" />
-    <PackageVersion Include="Microsoft.Web.Xdt" Version="3.0.0" />
+    <PackageVersion Include="Microsoft.Web.Xdt" Version="$(MicrosoftWebXdtPackageVersion)" />
     <PackageVersion Include="Moq" Version="4.18.1" />
     <PackageVersion Include="MSTest.TestAdapter" Version="2.2.10" />
     <PackageVersion Include="MSTest.TestFramework" Version="2.2.10" />

